### ISSUE TITLE
It is possible to create a continuous flow?

### ISSUE THREAD

**Issue by paolobolzoni** [taskflow/taskflow#66]
The situation I am describing is based on my use case, but it has simplified. So, please, do not focus on the details unless they are needed to answer.

I have a flow of data, from our point of view we can see it as a `volatile` variable I have to read regularly.
From the value of the variable I need to compute five functions: `f0`, `f1`, `f2`, `f3`,  and `f4`. Each takes longer than the previous and uses a input the output of the previous. `f0` starts with the `volatile` variable value.

This situation calls for the pipeline pattern, basically after five readings we are doing all the computations in parallel.

Playing around with **cppflow** (great library, by the way) I understood how to make _one_ loop, however this is not helpful as if there is only one reading then the computation is basically single threaded.

So my question is, it is possible to implement something like this? I would be happy to read the fine manual if the answer is there, just point me in the right direction.

**Comment by twhuang-uiuc** [comment]
Hi @paolobolzoni it seems to me this is a sequential flow where you have something like the following:
```cpp
volatile int a;
auto res1 = f0(a);
auto res2 = f1(res1);
auto res3 = f2(res2);
auto res4 = f3(res3);
auto res5 = f4(res4);
```
Please let me know if I misunderstood your problem.

**Comment by paolobolzoni** [comment]
I probably explain badly, that is indeed the flow.
But  the idea is to execute f0, f1, f2, f3, and f4 at the same time. Because I need to read from the volatile variable continuously.

So basically it should go like this
`f0`
`f0` `f1`
`f0` `f1` `f2`
`f0` `f1` `f2` `f3`
`f0` `f1` `f2` `f3` `f4`
`f0` `f1` `f2` `f3` `f4`
... and continues like this

The function call on the same line are executed simultaneously.

**Comment by twhuang-uiuc** [comment]
I believe you can create a task dependency graph for this. It should be something like the following:
```cpp
// create tasks
std::vector<std::vector<tf::Task>> tasks;
for(int i=0; i<layers; i=i+1) {
  for(int j=0; j<min(i, 5); j++) {
    tasks[i][j] = tf.silent_emplace([](){  /* add your functions */  });
  }
}

// add dependencies
for(int i=0; i<layers; i=i+1) {
  for(int j=0; j<min(i, 5); j++) {
    tasks[i-1][j-1].precede(tasks[i][j]);
    tasks[i][j-1].precede(tasks[i][j]);
  }
}
```

**Comment by paolobolzoni** [comment]
That idea will not work because I do not know the value of `layers`, the code will run for hours or maybe days, I do not know the number of times I will call `f0`. Even if I did it can be fairly large.


Perhaps the correct question is, can I execute a `tf:Task` more than once?

My idea was to create the dependency graph using `.precede` like this:

   f0 -> f1 -> f2 -> f3 -> f4

And call `f0` multiple times, since the `f0` output depends on an external state the final `f4` output will be different for each call.

Here is a reference to the pipeline pattern; perhaps it will help understand each other.
https://www.cise.ufl.edu/research/ParallelPatterns/PatternLanguage/AlgorithmStructure/Pipeline.htm

**Comment by twhuang-uiuc** [comment]
Hi @paolobolzoni we are working on a special feature called *framework* that allows users to execute a task dependency graph multiple times. We will keep you posted. 

I think based on your description, you are asking for a stateful execution. You want to execute `f0->f1->f2->f3->f4` for several times, correct?

**Comment by paolobolzoni** [comment]
Exactly, while it is not my use case I think a videogame-inspired flow shows the point very well.

The functions take the inpute from the previous one and give output to the next and are:

`read_player_input`
`compute_ai`
`compute_physics`
`compute_sound`
`compute_frame_and_show`

The first function input is unpredictable (just like a `volatile` variable), and the last function output is shown to the user (on the screen and via sounds).

Thinking low level, we can setup a producer/consumer chain of threads, each function is executed in a different thread.
So the `read_player_input` produces and send to the consumer `compute_ai`; similarly we have the other steps (`compute_ai` consumes the input from `read_player_input` and sends its results to `compute_physics`...); finally `compute_frame` fill up the screen and sound card buffers... somewhat.

The point I am trying to pass is that there is no reason for `read_player_input` to wait that the previous `compute_frame_and_show` is complete. When the game is running, the functions can work simultaneously. 
The `read_player_input` function will be called regularly to ensure the desired framerate, or as fast as possible if the hardware is not fast enough.


Introducing another example often introduces confusion, I hope in this case helps.

**Comment by twhuang-uiuc** [comment]
Yes, in fact, I did something similar to your example in my another VLSI timing analysis project using taskflow to avoid the synchronization. What I did is to linearize the sequences and construct a task dependency graph. Something like the following:
```
// assume only three layers
r1=read_player_input_1
c1_1=compute_ai_1
c2_1=compute_physics_1
c3_1=compute_sound_1
c4_1=compute_frame_and_show_1
tf.linearize(r1, c1_1, c2_1, c3_1, c4_1);

r2=read_player_input_2
c1_2=compute_ai_2
c2_2=compute_physics_2
c3_2=compute_sound_2
c4_2=compute_frame_and_show_2
tf.linearize(r2, c1_2, c2_2, c3_2, c4_2);

r3=read_player_input_3
c1_3=compute_ai_3
c2_3=compute_physics_3
c3_3=compute_sound_3
c4_3=compute_frame_and_show_3
tf.linearize(r3, c1_3, c2_3, c3_3, c4_3);

// now construct inter-layer dependencies
r1.precede(r2);
c1_1.precede(c1_2);
c2_1.precede(c2_2);
c3_1.precede(c3_2);
c4_1.precede(c4_2);

r2.precede(r3);
c1_2.precede(c1_3);
c2_2.precede(c2_3);
c3_2.precede(c3_3);
c4_2.precede(c4_3);
```
Is this something you can do? I recalled you said you don't know the number of the layers? Then how do you terminate the program?

**Comment by paolobolzoni** [comment]
Keeping the videogame analogy, the programs stops when the `read_player_input` function reads the keypress "Esc" or something similar. `read_player_input` will send as output to the rest of the functions a special value that basically tells them to halt. Whatever function started everything waits that everyone joins.

It is impossible to know when (or even _if_) it will happen, the player can go on for hours or stop after five minutes.


In my case, I am working in a simulation and the plan is making it halt when we get specifics ranges of values. Or we give up and press CTRL-C ...

**Comment by twhuang-uiuc** [comment]
Hi @paolobolzoni  thanks for bringing this up. We are working on a new feature called `framework` that might suit your purpose. It will come out very soon. Let me keep you posted and we can go from there to see if that helps.

**Comment by twhuang-uiuc** [comment]
@paolobolzoni , we have

**Comment by twhuang-uiuc** [comment]
@paolobolzoni , we have just released the Framework feature in the master branch. This allows users to reuse a task dependency graph and execute it multiple times. I think it might be useful for your workload here. Please refer to [Framework Tutorial](https://cpp-taskflow.github.io/cpp-taskflow/chapter7.html).

Please let use know what/how you think.

**Comment by paolobolzoni** [comment]
I am not sure I can use it, let us start from a slightly modified example from the documentation. A pipeline of three elements.

```
#include <taskflow/taskflow.hpp>
#include <unistd.h>

int main() {
tf::Framework f{};

int ac{}, bc{}, cc{};
auto [A, B, C] = f.emplace(
  [&ac] ()  { sleep(1); std::cout << "This is TaskA" << ++ac << "\n"; },
  [&bc] ()  { sleep(2); std::cout << "This is TaskB" << ++bc << "\n"; },
  [&cc] ()  { sleep(3); std::cout << "This is TaskC" << ++cc << "\n"; }
);

A.precede(B); 
B.precede(C); 

// Declare a taskflow object
tf::Taskflow taskflow{};

int cnt{}; taskflow.run_until(f, [&cnt] () { return (++cnt == 10); }); // my compiler did not like the mutable lambda
}
```

The output is see is:
```
This is TaskA1
This is TaskB1
This is TaskC1
This is TaskA2
This is TaskB2
This is TaskC2
This is TaskA3
This is TaskB3
This is TaskC3
This is TaskA4
...
```
Each execution of `f` requires six seconds. The idea of the pipeline is to have only the first execution to last six seconds and all the consecutive ones last only last three seconds. Because when the the execution number 1 is running, the execution number 2 can already start.

The output should be on the lines of (I am making the order on my head, caveat emptor):
```
This is TaskA1
This is TaskB1
This is TaskA2
This is TaskC1
This is TaskB2
This is TaskA3
This is TaskC2
This is TaskB3
This is TaskA4
This is TaskC3
...
```

It is well possible that I did not understand or that it is possible to workaround using a slightly more complex task graph. I have to try a bit.
About the latter, the task graph in the https://github.com/cpp-taskflow/cpp-taskflow/issues/66#issuecomment-462054838 may work.

**Comment by paolobolzoni** [comment]
As a matter of fact, using this code the timing is 12 seconds as expected. Because the first run uses 6 seconds, but the two next ones can overlap for three seconds on the previous one and, of course, 6+3+3=12.

So, I would say it works... but it is still a bit convoluted. How difficult would be to implement a `run_until` that restarts the execution without waiting that the previous one has finished?

```
#include <taskflow/taskflow.hpp>
#include <unistd.h>

int main() {
tf::Framework f{};

int ac{}, bc{}, cc{};
auto A = [&ac] ()  { sleep(1); std::cout << "This is TaskA" << ++ac << "\n"; };
auto B = [&bc] ()  { sleep(2); std::cout << "This is TaskB" << ++bc << "\n"; };
auto C = [&cc] ()  { sleep(3); std::cout << "This is TaskC" << ++cc << "\n"; };

auto S = f.emplace([](){});
auto [A1, B1, C1] = f.emplace(A,B,C);
auto [A2, B2, C2] = f.emplace(A,B,C);
auto [A3, B3, C3] = f.emplace(A,B,C);

S.precede(A1, A2, A3);

A1.precede(B1); 
B1.precede(C1); 

A2.precede(B2); 
B2.precede(C2); 

A3.precede(B3); 
B3.precede(C3); 

A1.precede(A2);
A2.precede(A3);
B1.precede(B2);
B2.precede(B3);
C1.precede(C2);
C2.precede(C3);

tf::Taskflow taskflow{};
taskflow.run_until(f, [&cc] () { return (cc == 3); }); }
```

**Comment by paolobolzoni** [comment]
The thumbs up is nice but a bit inscrutable, how difficult would it be to implement a `run_until` that restarts the execution without waiting that the previous one has finished?

**Comment by clin99** [comment]
Hi @paolobolzoni , I am looking into this and will keep you posted. Thank you for your patience.

**Comment by paolobolzoni** [comment]
Excellent news. Thanks for all the good work

Incidentally, I was wondering. Is there interest for pull requests about the documentation? Seeing the issue (#41) I think I can write how I passed the data in my flow. Of course, I understand if it is preferred to leave the documentation to the developers as the most experts.

**Comment by twhuang-uiuc** [comment]
@paolobolzoni Sure! You can also contribute an example to the folder `example/`. We will ack you in the project site :)

**Comment by GarrickLin** [comment]
Is there any elegant way to implement pipeline yet ? [Framework Tutorial](https://cpp-taskflow.github.io/cpp-taskflow/chapter7.html) cannot be visited now.

**Comment by paolobolzoni** [comment]
My examples have been moved in the `legacy` directory, I am not sure how useful they can still be. But have a look:

https://github.com/cpp-taskflow/cpp-taskflow/blob/master/legacy/20190514/example/get_best_dice.cpp
https://github.com/cpp-taskflow/cpp-taskflow/blob/master/legacy/20190514/example/get_best_dice.cpp

**Comment by tsung-wei-huang** [comment]
@paolobolzoni @GarrickLin The newest master branch has included a new feature of [pipeline scheduling framework](https://taskflow.github.io/taskflow/ParallelPipeline.html) that allows you to implement a parallel pipeline application. I believe this is what you want. Pipeline will be included in the next release v3.3.  Please have a try and let us know if you have any comments/suggestions.

Feel free to reopen this issue if necessary.

**Comment by lix19937** [comment]
see  https://github.com/taskflow/taskflow/issues/529  , is it support ? @tsung-wei-huang  thanks
